File: 1275.find-winner-on-a-tic-tac-toe-game.cpp
/*
 * @lc app=leetcode id=1275 lang=cpp
 *
 * [1275] Find Winner on a Tic Tac Toe Game
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  string tictactoe(vector<vector<int>>& moves) {
    char tiles[3][3];
    int size = moves.size();
    memset(tiles, 0x0, sizeof(tiles));
    for (int i = 0; i < size; i++) {
      tiles[moves[i][0]][moves[i][1]] = (i % 2) ? -1 : 1;
    }

    int sum = 0;
    for (int i = 0; i < 3; i++) {
      sum = tiles[i][0] + tiles[i][1] + tiles[i][2];
      if (sum == 3) {
        return "A";
      } else if (sum == -3) {
        return "B";
      }
      sum = tiles[0][i] + tiles[1][i] + tiles[2][i];
      if (sum == 3) {
        return "A";
      } else if (sum == -3) {
        return "B";
      }
    }
    sum = tiles[0][0] + tiles[1][1] + tiles[2][2];
    if (sum == 3) {
      return "A";
    } else if (sum == -3) {
      return "B";
    }
    sum = tiles[0][2] + tiles[1][1] + tiles[2][0];
    if (sum == 3) {
      return "A";
    } else if (sum == -3) {
      return "B";
    }

    for (int i = 0; i < 3; i++) {
      for (int j = 0; j < 3; j++) {
        if (tiles[i][j] == 0) {
          return "Pending";
        }
      }
    }

    return "Draw";
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<vector<int>> moves;
  moves = {{0, 0}, {2, 0}, {1, 1}, {2, 1}, {2, 2}};
  ASSERT_EQ(s.tictactoe(moves), "A");
  moves = {{0, 0}, {1, 1}, {0, 1}, {0, 2}, {1, 0}, {2, 0}};
  ASSERT_EQ(s.tictactoe(moves), "B");
  moves = {{0, 0}, {1, 1}, {2, 0}, {1, 0}, {1, 2},
           {2, 1}, {0, 1}, {0, 2}, {2, 2}};
  ASSERT_EQ(s.tictactoe(moves), "Draw");
  moves = {{0, 0}, {1, 1}};
  ASSERT_EQ(s.tictactoe(moves), "Pending");
  return 0;
}

File: 1275.find-winner-on-a-tic-tac-toe-game.rs
/*
 * @lc app=leetcode id=1275 lang=rust
 *
 * [1275] Find Winner on a Tic Tac Toe Game
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn tictactoe(moves: Vec<Vec<i32>>) -> String {
        let mut a = [0; 8];
        let mut b = [0; 8];
        for i in 0..moves.len() {
            let r = moves[i][0];
            let c = moves[i][1];
            if i % 2 == 0 {
                a[r as usize] += 1;
                a[(c + 3) as usize] += 1;
                if r == c {
                    a[6] += 1;
                }
                if r == (2 - c) {
                    a[7] += 1;
                }
            } else {
                b[r as usize] += 1;
                b[(c + 3) as usize] += 1;
                if r == c {
                    b[6] += 1;
                }
                if r == (2 - c) {
                    b[7] += 1;
                }
            }
        }
        for i in 0..8 {
            if a[i] == 3 {
                return "A".to_string();
            }
            if b[i] == 3 {
                return "B".to_string();
            }
        }
        if moves.len() == 9 {
            "Draw".to_string()
        } else {
            "Pending".to_string()
        }
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::tictactoe(vec![
            vec![0, 0],
            vec![2, 0],
            vec![1, 1],
            vec![2, 1],
            vec![2, 2]
        ]),
        "A".to_string()
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::tictactoe(vec![
            vec![0, 0],
            vec![1, 1],
            vec![0, 1],
            vec![0, 2],
            vec![1, 0],
            vec![2, 0]
        ]),
        "B".to_string()
    );
}

#[test]
fn test3() {
    assert_eq!(
        Solution::tictactoe(vec![
            vec![0, 0],
            vec![1, 1],
            vec![2, 0],
            vec![1, 0],
            vec![1, 2],
            vec![2, 1],
            vec![0, 1],
            vec![0, 2],
            vec![2, 2]
        ]),
        "Draw".to_string()
    );
}
